### PR TITLE
[offload] Add support for multiple streams and fix issue with integer inputs.

### PR DESCRIPTION
**Pull request by anj-s**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
This PR does the following:
- Removes the need to set requires_grad on all inputs(beginning and intermediate) and is the better long term solution whereby we use a dummy tensor to trigger the BW pass. (Thanks to Alban for helping with this tricky issue.)
- Perf improvement changes such as using multiple streams, pinning model shard memory and removing certain copies that allows overlap of copies. More changes coming soon.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: benchmarks/experimental/offload.py [hunk 1]
```diff
@@ -149,6 +150,10 @@ def train_epoch(args):
 
 
 def train(model_config, model, benchmark_config, model_specs, args):
+    device = torch.device("cuda")
+    torch.cuda.set_device(0)
+    torch.manual_seed(5)
```
Comment (blefaudeux) [inline benchmarks/experimental/offload.py:155]: jus curious, why the 5 ?

File: benchmarks/experimental/offload.py [hunk 2]
```diff
@@ -208,6 +215,8 @@ def get_batch(source):
             total_tokens_per_log_interval = 0
             total_loss = 0
             start_time = time.time()
+        prof.export_chrome_trace("/tmp/offload_prof")
```
Comment (blefaudeux) [inline benchmarks/experimental/offload.py:218]: pretty practical :)

File: fairscale/experimental/nn/offload.py [hunk 3]
```diff
@@ -112,20 +113,20 @@ def to_device(self) -> None:
         self.model_shard.to(device=self.device, non_blocking=True)
 
     def forward_load(self, non_blocking: bool = True) -> None:
-        with torch.cuda.stream(self.cuda_stream):
+        with torch.cuda.stream(self._cpu_to_gpu_stream):
```
Comment (blefaudeux) [inline fairscale/experimental/nn/offload.py:116]: nice

File: fairscale/experimental/nn/offload.py [hunk 4]
```diff
@@ -164,10 +165,11 @@ def forward(ctx: Any, inputs: Any, model_instance: Any) -> Any:
         model_instance._activations = [inputs]
         # Enumerate through layer shards and apply activations from the previous shard.
         for index, layer_shard in enumerate(model_instance.model_slices):
-            # Bring in the current activations onto the device.
-            model_instance._activations[index] = tuple([a.cuda() for a in list(model_instance._activations[index])])
             # Bring in the current layer shard onto the device.
             layer_shard.forward_load()
+            # Bring in the current activations onto the device.
```
Comment (blefaudeux) [inline fairscale/experimental/nn/offload.py:170]: any logic as of why this is flipped (layer/activation load) when compared to before ?

**Review by blefaudeux — APPROVED** [review]
LGTM, a couple of places where I could do with some explanations, but overall makes sense

File: benchmarks/experimental/offload.py (hunk 1 above)
Comment (anj-s) [reply benchmarks/experimental/offload.py]: Just a number for me to compare weights. Removing this in a follow up PR where I add golden numbers for the benchmarks.

File: fairscale/experimental/nn/offload.py (hunk 4 above)
Comment (anj-s) [reply fairscale/experimental/nn/offload.py]: I was trying out some stuff and flipped the order for readability. Flipped it back to avoid confusion.

**Pull request merged**